File: tools/8.5/fetch-sources.py
"""Fetch the exact PHP-8.5 revision audited by this repository into a local directory."""
import argparse
import base64
import json
import urllib.request
from pathlib import Path

parser = argparse.ArgumentParser()
parser.add_argument('destination', type=Path)
args = parser.parse_args()
args.destination.mkdir(parents=True, exist_ok=True)
revision = '7a4c62795365ed6a97a0184c96375b9fb4d53b1e'
for name in ['zend_language_parser.y', 'zend_language_scanner.l', 'zend_compile.c']:
    url = f'https://api.github.com/repos/php/php-src/contents/Zend/{name}?ref={revision}'
    with urllib.request.urlopen(url) as response:
        data = json.load(response)
    (args.destination / name).write_bytes(base64.b64decode(data['content']))
print(revision)

File: tools/8.5/fetch-sources.php
<?php
declare(strict_types=1);
require __DIR__ . '/../lib/bootstrap.php';
use PhpGrammar\Tools\Support as S;
$args = S::args(min: 1, max: 1);
$destination = $args['positionals'][0];
if (!is_dir($destination)) {
    mkdir($destination, 0777, true);
}
foreach (['zend_language_parser.y', 'zend_language_scanner.l', 'zend_compile.c'] as $name) {
    $url = 'https://api.github.com/repos/php/php-src/contents/Zend/' . $name . '?ref=' . S::PIN;
    $data = json_decode(S::download($url), true, 512, JSON_THROW_ON_ERROR);
    $bytes = base64_decode($data['content'], true);
    if ($bytes === false) {
        throw new RuntimeException('Invalid source encoding: ' . $name);
    }
    file_put_contents($destination . '/' . $name, $bytes);
}
echo S::PIN . "\n";
